src/Storage/Notmuch.hs

```diff
@@ -78,24 +85,31 @@ mailFilepath m dbpath =
 setNotmuchMailTags
   :: (MonadError Error m, MonadIO m)
   => FilePath
+  -> [T.Text]
+  -> ([T.Text] -> NotmuchMail -> NotmuchMail)
   -> NotmuchMail
   -> m NotmuchMail
-setNotmuchMailTags dbpath m = do
-  nmtags <- toNotmuchTags (view mailTags m)
-  bracketT (Notmuch.databaseOpen dbpath) Notmuch.databaseDestroy (tagsToMessage nmtags (view mailId m))
-  pure m
+setNotmuchMailTags dbpath tgs op m = do
+  let m' = op tgs m
+  nmtags <- toNotmuchTags (view mailTags m')
+  bracketT (Notmuch.databaseOpen dbpath) Notmuch.databaseDestroy (tagsToMessage nmtags (view mailId m'))
+  pure m'
 
```

Comment:
We want to decompose *get all messages for thread* from *do these tag operations to each of these messages*.

```haskell

data TagOp = TagRemove Tag | TagAdd Tag

messageSetTagModify
  :: (Traversable t, MonadError e m, MonadIO m)
  => FilePath {- database -} 
  -> [TagOp]
  -> t NotmuchMail             -- set of messages
  -> m (t NotmuchMail)

-- and there'll a separate operation for getting the messages to operate on
```

Reply:
So this is working out fine. I like it.

The only trouble I've ran into now is the disparity in the UI. For the case that there are many mails selected updating is straight forward. The outlier is the thread here and I'm not sure how to go about it. If I stick with a possible traversable of `NotmuchMail` I could figure out the thread and reload it in the list. I could also just pass the current selected item around, which seems to spoil this, because the whole idea is that we could have multiple selected. 

Having said that (or just written down), the traversable of `NotmuchMail` is actually whatever item I have selected in the UI. Possibly multiple in the future, including threads.

Perhaps time to sleep about it.